cachi2/core/package_managers/npm.py

```diff
@@ -0,0 +1,226 @@
+        key for registry deps and the "version" key for non-registry deps.
+
+        For v2+ package-lock.json packages, this will be the "resolved" key.
+        """
+        key = "resolved" if "resolved" in self._package_dict else "version"
+        self._package_dict[key] = resolved_url
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Package):
+            return (
+                self.name == other.name
+                and self.path == other.path
+                and self._package_dict == other._package_dict
+            )
+        return False
+
+
+class PackageLock:
+    """A npm package-lock.json file."""
+
+    def __init__(
+        self, lockfile_path: RootedPath, output_dir: RootedPath, lockfile_data: Dict[str, Any]
+    ) -> None:
+        """Initialize a PackageLock."""
+        self._lockfile_path = lockfile_path
+        self._output_dir = output_dir
+        self._lockfile_data = lockfile_data
+        self._packages = self._get_packages()
+        self._dependencies = self._get_dependencies()
+
+    @property
+    def lockfile_version(self) -> int:
+        """Get the lockfileVersion from package-lock.json data."""
+        return self._lockfile_data["lockfileVersion"]
+
+    @property
+    def main_package(self) -> Dict[str, str]:
+        """Return a dict with sbom component data for the main package."""
+        return {"name": self._lockfile_data["name"], "version": self._lockfile_data["version"]}
+
+    @classmethod
+    def from_file(cls, lockfile_path: RootedPath, output_dir: RootedPath) -> "PackageLock":
+        """Create a PackageLock from a package-lock.json file."""
+        with lockfile_path.path.open("r") as f:
+            lockfile_data = json.load(f)
+
+        lockfile_version = lockfile_data.get("lockfileVersion")
+        if lockfile_version not in (1, 2, 3):
+            raise PackageRejected(
+                f"lockfileVersion {lockfile_version} from {lockfile_path} is not supported",
+                solution="Please use a supported lockfileVersion, which are versions 1, 2, and 3",
+            )
+
+        return cls(lockfile_path, output_dir, lockfile_data)
+
+    def get_project_file(self) -> ProjectFile:
+        """Return a ProjectFile for the npm package-lock.json data."""
+        return ProjectFile(
+            abspath=self._lockfile_path.path.resolve(),
+            template=json.dumps(self._lockfile_data, indent=2) + "\n",
+        )
+
+    def _get_packages(self) -> List[Package]:
+        """Return a flat list of Packages from a v2+ package-lock.json file.
+
+        Use the "packages" key in the lockfile.
+        """
+
+        def get_package_name_from_path(package_path: str) -> str:
+            """Get the package name from the path in v2+ package-lock.json file."""
+            parent = Path(package_path).parents[0]
+            if parent.name.startswith("@"):
+                return os.path.join(parent.name, Path(package_path).name)
```

Comment:
I'd recommend against `os.path`, Path is just a better API

And in this case, I think we always want forward slashes in the name so using platform-specific path APIs kinda sends the wrong message. `Path.as_posix` or just a string join would IMO be more appropriate